File: eclipse/plugin/src/test/java/com/adityakumar/engine/SorterTest.java
package com.adityakumar.engine;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class SorterTest {

    // ---- numeric sort (the bug: was lexicographic, must be numeric) ----

    /** Core regression: ["10","9","100"] must sort to [9,10,100], NOT [10,100,9]. */
    @Test
    void numericAscNotLexicographic() {
        var result = Sorter.sortLines(
                List.of("10", "9", "100"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.ASC, false, true, null));
        assertEquals(List.of("9", "10", "100"), result,
                "NUMERIC ASC must order by value, not lexicographically");
    }

    @Test
    void numericDescNotLexicographic() {
        var result = Sorter.sortLines(
                List.of("10", "9", "100"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.DESC, false, true, null));
        assertEquals(List.of("100", "10", "9"), result);
    }

    @Test
    void nonNumericKeysOrderedAfterNumeric() {
        // Non-parseable keys must appear AFTER all numeric keys in ASC order.
        var result = Sorter.sortLines(
                List.of("abc", "5", "10", "1"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.ASC, false, true, null));
        assertEquals(List.of("1", "5", "10", "abc"), result);
    }

    @Test
    void floatingPointNumericSort() {
        var result = Sorter.sortLines(
                List.of("3.14", "1.0", "2.718"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.ASC, false, true, null));
        assertEquals(List.of("1.0", "2.718", "3.14"), result);
    }

    @Test
    void negativeNumbersNumericSort() {
        var result = Sorter.sortLines(
                List.of("5", "-3", "0", "10"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.ASC, false, true, null));
        assertEquals(List.of("-3", "0", "5", "10"), result);
    }

    // ---- alphabetical sort ----

    @Test
    void alphabeticalAscCaseInsensitive() {
        var result = Sorter.sortLines(
                List.of("banana", "Apple", "cherry"),
                new SortOptions(SortOptions.Mode.ALPHABETICAL, SortOptions.Direction.ASC, true, true, null));
        assertEquals(List.of("Apple", "banana", "cherry"), result);
    }

    @Test
    void alphabeticalDesc() {
        var result = Sorter.sortLines(
                List.of("a", "c", "b"),
                new SortOptions(SortOptions.Mode.ALPHABETICAL, SortOptions.Direction.DESC, false, true, null));
        assertEquals(List.of("c", "b", "a"), result);
    }

    // ---- trim interacts with numeric ----

    @Test
    void numericWithLeadingSpaceTrimmed() {
        // "  10" with trim=true should be treated as 10, not "  10" (which is NaN).
        var result = Sorter.sortLines(
                List.of("  10", " 9", "100"),
                new SortOptions(SortOptions.Mode.NUMERIC, SortOptions.Direction.ASC, false, true, null));
        assertEquals(List.of(" 9", "  10", "100"), result);
    }
}

File: jetbrains/src/test/kotlin/engine/SorterTest.kt
package com.adityakumar.engine

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class SorterTest {
    @Test fun `alphabetical ascending`() {
        val r = Sorter.sortLines(listOf("c","a","b"), SortOptions())
        assertEquals(listOf("a","b","c"), r)
    }
    @Test fun `alphabetical descending`() {
        val r = Sorter.sortLines(listOf("a","c","b"), SortOptions(direction=SortDirection.DESC))
        assertEquals(listOf("c","b","a"), r)
    }
    @Test fun `numeric sort`() {
        val r = Sorter.sortLines(listOf("10","2","1"), SortOptions(mode=SortMode.NUMERIC))
        assertEquals(listOf("1","2","10"), r)
    }
    @Test fun `column extraction`() {
        val col = ColumnSpec(",", 2)
        val r = Sorter.sortLines(listOf("b,2","a,1","c,3"), SortOptions(column=col))
        assertEquals(listOf("a,1","b,2","c,3"), r)
    }
    @Test fun `case insensitive`() {
        val r = Sorter.sortLines(listOf("B","a","C"), SortOptions(caseInsensitive=true))
        assertEquals(listOf("a","B","C"), r)
    }
}
